subxt/src/client/lightclient/background.rs

```diff
@@ -0,0 +1,446 @@
+        params: String,
+        /// Channel used to send back the subscription ID if successful.
+        sub_id: oneshot::Sender<MethodResponse>,
+        /// Channel used to send back the notifcations.
+        sender: mpsc::Sender<Box<RawValue>>,
+    },
+}
+
+/// Background task data.
+pub struct BackgroundTask {
+    /// Smoldot light client implementation that leverages the exposed platform.
+    client: smoldot_light::Client<Platform>,
+    /// The ID of the chain used to identify the chain protocol (ie. substrate).
+    ///
+    /// Note: A single chain is supported for a client. This aligns with the subxt's
+    /// vision of the Client.
+    chain_id: ChainId,
+    /// Unique ID for RPC calls.
+    request_id: usize,
+    /// Map the request ID of a RPC method to the frontend `Sender`.
+    requests: HashMap<SubscriptionId, oneshot::Sender<MethodResponse>>,
+    /// Subscription calls first need to make a plain RPC method
+    /// request to obtain the subscription ID.
+    ///
+    /// The RPC method request is made in the background and the response should
+    /// not be sent back to the user.
+    /// Map the request ID of a RPC method to the frontend `Sender`.
+    id_to_subscription:
+        HashMap<SubscriptionId, (oneshot::Sender<MethodResponse>, mpsc::Sender<Box<RawValue>>)>,
+    /// Map the subscription ID to the frontend `Sender`.
+    subscriptions: HashMap<SubscriptionId, mpsc::Sender<Box<RawValue>>>,
+}
+
+impl BackgroundTask {
+    /// Constructs a new [`BackgroundTask`].
+    pub fn new(client: smoldot_light::Client<Platform>, chain_id: ChainId) -> BackgroundTask {
+        BackgroundTask {
+            client,
+            chain_id,
+            request_id: 1,
+            requests: Default::default(),
+            id_to_subscription: Default::default(),
+            subscriptions: Default::default(),
+        }
+    }
+
+    /// Fetch and increment the request ID.
+    fn next_id(&mut self) -> usize {
+        let next = self.request_id;
+        self.request_id = self.request_id.wrapping_add(1);
+        next
+    }
+
+    /// Handle the registration messages received from the user.
+    async fn handle_requests(&mut self, message: FromSubxt) {
+        match message {
+            FromSubxt::Request {
+                method,
+                params,
+                sender,
+            } => {
+                let id = self.next_id();
+                let request = format!(
+                    r#"{{"jsonrpc":"2.0","id":"{}", "method":"{}","params":{}}}"#,
+                    id, method, params
+                );
+
+                self.requests.insert(id, sender);
+
+                let result = self.client.json_rpc_request(request, self.chain_id);
+                if let Err(err) = result {
+                    tracing::warn!(
+                        target: LOG_TARGET,
+                        "Cannot send RPC request to lightclient {:?}",
+                        err.to_string()
+                    );
```

Comment:
I just want to point out that this can realistically happen. If a `TooBusy` error is returned, you're supposed to back-pressure the requests sender.

I understand that from your point of view it's not very convenient, so I've opened https://github.com/smol-dot/smoldot/issues/804 to clarify this thing.

Reply:
To clarify: there will always be an error here. But maybe I can provide some configuration option that guarantees a minimum buffer size, so that you have a runtime guarantee that it won't happen.

Reply:
Sorry I was a bit confused this morning since I was working on something else in parallel.

What you should do is pass `max_pending_requests: u32::max_value()` and `max_subscriptions: u32::max_value()` when adding the chain. There's no need to make these parameters configurable.

These parameters exist to make sure that the memory usage of smoldot is bounded in situations where the JSON-RPC client is potentially malicious and can spam you with requests. This isn't relevant here, since you are the JSON-RPC client.